src/mpid/ch4/include/mpidpre.h

```diff
@@ -227,7 +227,31 @@ typedef struct MPIDI_prequest {
     MPI_Datatype datatype;
 } MPIDI_prequest_t;
 
+/* Structures for partitioned pt2pt request */
+typedef struct MPIDIG_part_sreq {
+    MPIR_cc_t ready_cntr;
+} MPIDIG_part_sreq_t;
+
+typedef struct MPIDIG_part_rreq {
+    MPI_Aint sdata_size;        /* size of entire send data */
+
+    /* fields for posted/unexp list */
+    MPIR_Request *request;      /* pointer of part_req */
+    struct MPIDIG_part_rreq *prev, *next;
+} MPIDIG_part_rreq_t;
+
+typedef struct MPIDIG_part_request {
+    MPL_atomic_int_t status;    /* see MPIDIG_PART_REQ_INC_FETCH_STATUS */
+    MPIR_Request *peer_req_ptr;
+    union {
+        MPIDIG_part_sreq_t sreq;
+        MPIDIG_part_rreq_t rreq;
+    } u;
+} MPIDIG_part_request_t;
+
 typedef struct MPIDI_part_request {
+    MPIDIG_part_request_t am;
+
```

Comment:
Could you remind me again why this `am` cannot be a part of the AM union as the other AM requests (e.g. `(req)->dev.ch4.am.part`)?

Reply:
This `am` is in `(req)->dev.ch4.part_req` which is union with `(req)->dev.ch4.am`. 

`(req)->dev.ch4.part_req` is used by the "persistent" request created at `psend|precv_init`, whereas `(req)->dev.ch4.am` is used by the "one-time" request created at each AM-based data transfer.

We also define an extended type in `(req)->dev.ch4.am` for partitioned AM data transfer, see `MPIDIG_REQUEST(rreq, req->part_am_req)`. It stores the pointer of `(req)->dev.ch4.part_req` so that we can inactivate `part_req` when the "one-time" AM request completes (see `part_send_data_target_cmpl_cb` and `MPIDIG_part_send_data_origin_cb`)